server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1280,10 +1299,16 @@ protected InputStream openSlice(long slice) throws IOException {
                         }
                     }
                 }
-            }.restore(snapshotFiles, store);
-        } catch (Exception e) {
-            throw new IndexShardRestoreFailedException(shardId, "failed to restore snapshot [" + snapshotId + "]", e);
-        }
+            }.restore(snapshotFiles, store, l);
+        }));
+    }
+
+    private static ActionListener<Void> fileQueueListener(BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> files, int workers,
+                                                          ActionListener<Collection<Void>> listener) {
+        return ActionListener.delegateResponse(new GroupedActionListener<>(listener, workers), (l, e) -> {
+            files.clear(); // Stop uploading the remaining files if we run into any exception
```

Comment:
For extra safety can we call `onFailure()` in a finally block?
```java
try {
   files.clear(); // Stop uploading the remaining files if we run into any exception
} finally {
   l.onFailure(e);
}
```

Reply:
I shouldn't be necessary I think. If we run into an exception/throwable here the node is probably done for anyway? (the only way I see us getting here is some OOM, otherwise `clear()` should never throw on a healthy JVM?)